File: trunk/geopdes_fluid/inst/operators/op_f_curlv_2d.m
% OP_F_CURLV_2D: assemble the right-hand side vector r = [r(i)], with  r(i) = (f, curl v_i).
%
%   mat = op_f_curlv_2d (spv, msh, coeff);
%
% INPUT:
%     
%   spv:   structure representing the function space (see sp_bspline_2d_phys)
%   msh:   structure containing the domain partition and the quadrature rule (see msh_push_forward_2d)
%   coeff: source function evaluated at the quadrature points
%
% OUTPUT:
%
%   mat: assembled right-hand side
% 
% Copyright (C) 2009, 2010 Carlo de Falco
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.

%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.


function mat = op_f_curlv_2d (spv, msh, coeff)
  
 mat   = zeros(spv.ndof, 1);
 ndir = size (msh.quad_nodes, 1);

 shpv(1,:,:,:) =  spv.shape_function_gradients(2,:,:,:);
 shpv(2,:,:,:) = -spv.shape_function_gradients(1,:,:,:);

 coeff = reshape (coeff, ndir, msh.nqn, msh.nel);
 for iel = 1:msh.nel
   if (all (msh.jacdet(:,iel)))
     mat_loc = zeros (spv.nsh(iel), 1);
     for idof = 1:spv.nsh(iel)
       ishp = reshape(shpv(:,:,idof,iel), ndir, []);
       %        for inode = 1:msh.nqn
       mat_loc(idof) = mat_loc(idof) + ...
           sum (msh.jacdet(:, iel) .* msh.quad_weights(:, iel) .* ...
                sum(ishp .* coeff(:, :, iel), 1).');
       %        end  
     end
     mat(spv.connectivity(1:spv.nsh(iel), iel)) = mat(spv.connectivity(1:spv.nsh(iel), iel)) + mat_loc; 
   else
     warning ('geopdes:jacdet_zero_at_quad_node', 'op_f_curlv: singular map in element number %d', iel)
   end
 end
 
end

File: trunk/geopdes_fluid/src/op_f_curlv_2d.cc
/* Copyright (C) 2009 Carlo de Falco

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/


#include "geopdes.h"

DEFUN_DLD(op_f_curlv_2d, args, nargout,"")
{
  
  octave_value_list retval;

  geopdes_mesh  msh (args(1).map_value ());
  geopdes_space sp  (args(0).map_value (), msh);
  dim_vector    idx (2, msh.nqn (), msh.nel ());
  NDArray       coeff = args(2).array_value ().reshape (idx);

  if (!error_state)
    {

      ColumnVector mat (sp.ndof (), 0.0);

#pragma omp parallel default (none) shared (msh, sp, idx, mat, coeff)
      {
        double local_contribution;

#pragma omp for
        for (octave_idx_type iel=0; iel < msh.nel (); iel++) 
          if (msh.area (iel) > 0.0)
            {
              for (octave_idx_type idof(0); idof < sp.nsh (iel); idof++) 
                {
                  for ( octave_idx_type inode(0); inode < msh.nqn (); inode++) 
                    {
                      if (msh.weights (inode, iel) > 0.0)
                        {
                          double s = 0.0;
                          for (octave_idx_type icmp(0); icmp < 2; icmp++)
                            s += pow (-1.0, icmp) * sp.shape_function_gradients (0, 1-icmp, inode, idof, iel) * coeff (icmp, inode, iel);
                          local_contribution = msh.jacdet  (inode, iel) * msh.weights (inode, iel) * s;
#pragma omp critical
                          mat(sp.connectivity (idof, iel) - 1) += local_contribution;			  
                        }  
                    } // end for inode
                } // end for idof
            } else {
#pragma omp critical
            {warning_with_id ("geopdes:zero_measure_element", "op_f_v: element %d has 0 area (or volume)", iel);}
          }  // end for iel, if area > 0
      }  // end for parallel region
      retval(0) = octave_value (mat);
    } // end if !error_state
  return retval;
}

